messaging/src/main/java/org/axonframework/eventhandling/deadletter/DeadLetteringEventHandlerInvoker.java

```diff
@@ -116,33 +126,75 @@ public void handle(@Nonnull EventMessage<?> message, @Nonnull Segment segment) t
-        if (queue.enqueueIfPresent(sequenceIdentifier, () -> new GenericDeadLetter<>(sequenceIdentifier, message))) {
+        boolean skipIfPresentCheck = skipIfPresentCheck(message, sequenceIdentifier, segment);
+        if (!skipIfPresentCheck && queue.enqueueIfPresent(
+                sequenceIdentifier,
+                () -> new GenericDeadLetter<>(sequenceIdentifier, message)
+        )) {
             if (logger.isInfoEnabled()) {
                 logger.info("Event with id [{}] is added to the dead-letter queue "
                                     + "since its queue id [{}] is already present.",
                             message.getIdentifier(), sequenceIdentifier);
             }
+            markPresentInDLQ(sequenceIdentifier, segment);
         } else {
-            if (logger.isTraceEnabled()) {
-                logger.trace("Event [{}] with queue id [{}] is not present in the dead-letter queue."
-                                     + "Handle operation is delegated to the wrapped EventHandlerInvoker.",
-                             message, sequenceIdentifier);
+            if (!skipIfPresentCheck) {
+                markNotPresentInDLQ(sequenceIdentifier, segment);
             }
-            try {
-                super.invokeHandlers(message);
-            } catch (Exception e) {
-                DeadLetter<EventMessage<?>> letter = new GenericDeadLetter<>(sequenceIdentifier, message, e);
-                EnqueueDecision<EventMessage<?>> decision = enqueuePolicy.decide(letter, e);
-                if (decision.shouldEnqueue()) {
-                    Throwable cause = decision.enqueueCause().orElse(null);
-                    queue.enqueue(sequenceIdentifier, decision.withDiagnostics(letter.withCause(cause)));
-                } else if (logger.isInfoEnabled()) {
-                    logger.info("The enqueue policy decided not to dead letter event [{}].", message.getIdentifier());
-                }
+            invokeHandlers(message, segment, sequenceIdentifier);
+        }
+    }
+
+    private void invokeHandlers(@Nonnull EventMessage<?> message, @Nonnull Segment segment, Object sequenceIdentifier) {
+        if (logger.isTraceEnabled()) {
+            logger.trace("Event [{}] with queue id [{}] is not present in the dead-letter queue."
+                                 + "Handle operation is delegated to the wrapped EventHandlerInvoker.",
+                         message, sequenceIdentifier);
+        }
+        try {
+            super.invokeHandlers(message);
+        } catch (Exception e) {
+            DeadLetter<EventMessage<?>> letter = new GenericDeadLetter<>(sequenceIdentifier, message, e);
+            EnqueueDecision<EventMessage<?>> decision = enqueuePolicy.decide(letter, e);
+            if (decision.shouldEnqueue()) {
+                Throwable cause = decision.enqueueCause().orElse(null);
+                markPresentInDLQ(sequenceIdentifier, segment);
```

Comment:
I think that's much less worse than the alternative, as we will call the `enqueueIfPresent` next time, and if it's not in the dlq, it's fine. But if we switch them, and there is some error preventing the `markEnqueued` but it did store the event in the dlq, it will not put others belonging to the same sequence there, because the `enqueueIfPresent` will be skipped.